src/libopensc/asn1.c

```diff
@@ -89,9 +89,10 @@ int sc_asn1_read_tag(const u8 ** buf, size_t buflen, unsigned int *cla_out,
 			 * unsigned int */
 			return SC_ERROR_INVALID_ASN1_OBJECT;
 
+	} else {
+		if (left == 0)
+			return SC_ERROR_INVALID_ASN1_OBJECT;
 	}
-	if (left == 0)
-		return SC_ERROR_INVALID_ASN1_OBJECT;
 	/* parse length byte(s) */
 	len = *p & 0x7f;
 	if (*p++ & 0x80) {
```

Comment:
@frankmorgner 
"if (left == 0)" in your updated version will not work,
because of  " if (left < 2) return ...tt" few lines upper .

Better to remove it completely:
````
--- a/src/libopensc/asn1.c
+++ b/src/libopensc/asn1.c
@@ -87,14 +87,13 @@ int sc_asn1_read_tag(const u8 ** buf, size_t buflen, unsigned int *cla_out,
                if (left == 0 || n == 0)
                        /* either an invalid tag or it doesn't fit in
                         * unsigned int */
                        return SC_ERROR_INVALID_ASN1_OBJECT;
 
        }
-       if (left == 0)
-               return SC_ERROR_INVALID_ASN1_OBJECT;
+
        /* parse length byte(s) */
        len = *p & 0x7f;
        if (*p++ & 0x80) {
                unsigned int a = 0;
                if (len > 4 || len > left)
                        return SC_ERROR_INVALID_ASN1_OBJECT;
````